### PR TITLE
Switch to admission v1

### PR DESCRIPTION
**Pull request by jcaamano**
Switch to admission v1 from v1beta1. This drops support for k8s <= 1.15.

### CODE REVIEW

**Comment by jcaamano** [comment]
e2e is failing with

```
vagrant@devmaster:~/dev/k8snetworkplumbingwg/network-resources-injector> kubectl -n kube-system logs network-resources-injector installer
I0421 11:46:21.131894       1 main.go:28] starting webhook installation
I0421 11:46:21.132856       1 installer.go:45] generating Certificate Signing Request
2021/04/21 11:46:21 [INFO] received CSR
2021/04/21 11:46:21 [INFO] generating key: rsa-3072
2021/04/21 11:46:21 [INFO] encoded CSR
I0421 11:46:21.568517       1 installer.go:255] raw CSR and private key successfully created
I0421 11:46:21.574231       1 installer.go:62] CSR network-resources-injector-csr already exists, removing it first
I0421 11:46:21.578345       1 installer.go:66] creating new CSR network-resources-injector-csr
I0421 11:46:21.597046       1 installer.go:80] CSR pushed to the Kubernetes API
I0421 11:46:21.633650       1 installer.go:96] certificate approval sent
I0421 11:46:21.633666       1 installer.go:102] waiting for the signed certificate to be issued...
I0421 11:46:22.638622       1 installer.go:262] signed certificate successfully obtained
F0421 11:46:22.638657       1 installer.go:266] error writing certificate and key to files: open /etc/tls/tls.crt: permission denied
goroutine 1 [running]:
github.com/golang/glog.stacks(0xc0002ba600, 0xc000174280, 0x85, 0x9d)
	/usr/src/network-resources-injector/.gopath/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:769 +0xb8
github.com/golang/glog.(*loggingT).output(0x1ed3f20, 0xc000000003, 0xc00027db90, 0x1e42dc8, 0xc, 0x10a, 0x0)
	/usr/src/network-resources-injector/.gopath/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:720 +0x372
github.com/golang/glog.(*loggingT).printf(0x1ed3f20, 0x3, 0x13efa2f, 0x2e, 0xc000391e90, 0x1, 0x1)
	/usr/src/network-resources-injector/.gopath/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:655 +0x14b
github.com/golang/glog.Fatalf(...)
	/usr/src/network-resources-injector/.gopath/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:1148
github.com/k8snetworkplumbingwg/network-resources-injector/pkg/installer.Install(0x7ffe47b04bda, 0xb, 0x7ffe47b04bb4, 0x1a)
	/usr/src/network-resources-injector/pkg/installer/installer.go:266 +0x4c7
main.main()
	/usr/src/network-resources-injector/cmd/installer/main.go:29 +0x1ba
```

But no idea how this relates to the changes on this PR. Any ideas/hints?

**Comment by zshi-redhat** [comment]
> But no idea how this relates to the changes on this PR. Any ideas/hints?

Could it be that the container disk backing the emptyDir in NRI installer initContainer doesn't allow writing?
Perhaps changing the `emptyDir.medium` to `Memory` and see if it helps?

**Comment by jcaamano** [comment]
> > But no idea how this relates to the changes on this PR. Any ideas/hints?
> 
> Could it be that the container disk backing the emptyDir in NRI installer initContainer doesn't allow writing?
> Perhaps changing the `emptyDir.medium` to `Memory` and see if it helps?

Yes, I did try that and no dice :/

**Comment by jcaamano** [comment]
> > > But no idea how this relates to the changes on this PR. Any ideas/hints?
> > 
> > 
> > Could it be that the container disk backing the emptyDir in NRI installer initContainer doesn't allow writing?
> > Perhaps changing the `emptyDir.medium` to `Memory` and see if it helps?
> 
> Yes, I did try that and no dice :/

And if I go to HEAD~1 it works so it looks like is some change I did but I just don't see it.

**Review by martinkennelly — COMMENTED** [review]
@Eoghan1232 is looking into why this isn't working.

**Comment by Eoghan1232** [comment]
arv1 "k8s.io/api/admissionregistration/v1"

From what I've gathered so far, the installer.go file is breaking the PR. 
When you put the import back to "arv1beta1 : "k8s.io/api/admissionregistration/v1beta1" and change the clientset methods back to : clientset.AdmissionregistrationV1beta1() ....  the PR passes. 

Still looking into it, just posting an update so far.

**Comment by Eoghan1232** [comment]
deployments/server.yaml

Seems to be a bad image being pulled: it's trying to pull from docker.io as default.
When updated image to:
ghcr.io/k8snetworkplumbingwg/network-resources-injector:latest
PR passed.

The docker image seems to be 7 months old also? From docker.io... 
Is this a further issue?

**Comment by Eoghan1232** [comment]
It seems like there are is an issue with the volume mount permission with the newer K8s api changes.
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
The securityContext is now at pod level instead of at container level as done in the NRI server.yaml

Best thing to do is deploy the NRI and exec in and check the users on both and see if there is a mismatch.

I did a quick test, setting up a pod with 2 containers, and had them communicate with each other using the same volume mounted and securityContext setup.
I could see all process were done by the user I specified.
```
securityContext:
    runAsUser: 1000
```
I'll keep digging for now. 

@jcaamano Are you still looking into this?

File: pkg/installer/installer.go
Comment (jcaamano) [reply pkg/installer/installer.go]: This is now configurable so I guess it is ok that continues to be.

**Comment by jcaamano** [comment]
> It seems like there are is an issue with the volume mount permission with the newer K8s api changes.
> https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
> The securityContext is now at pod level instead of at container level as done in the NRI server.yaml
> 
> I did a quick test, setting up a pod with 2 containers, and had them communicate with each other using the same volume mounted and securityContext setup.
> I could see all process were done by the user I specified.
> 
> ```
> securityContext:
>     runAsUser: 1000
> ```
> 
> The containers could communicate fine, so it should work.
> 
> I'll keep digging for now.
> 
> @jcaamano Are you still looking into this?

I reproduced again, changed the command to a shell script with additional debug and then it did not happen any longer and  allowed me to realize sideEffects was missing from the webhook. So I went back to the code to see if the error creating the webhook was misinterpreted as the error writing the certificate, and I don't see it. I am not complaining.

I am using a different email so I guess it needs approval to run CI. cc @zshi-redhat @martinkennelly

**Comment by zshi-redhat** [comment]
@Eoghan1232 @martinkennelly would you mind take another look?

**Review by zshi-redhat — APPROVED** [review]

**Comment by Eoghan1232** [comment]
> > It seems like there are is an issue with the volume mount permission with the newer K8s api changes.
> > https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
> > The securityContext is now at pod level instead of at container level as done in the NRI server.yaml
> > I did a quick test, setting up a pod with 2 containers, and had them communicate with each other using the same volume mounted and securityContext setup.
> > I could see all process were done by the user I specified.
> > ```
> > securityContext:
> >     runAsUser: 1000
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The containers could communicate fine, so it should work.
> > I'll keep digging for now.
> > @jcaamano Are you still looking into this?
> 
> I reproduced again, changed the command to a shell script with additional debug and then it did not happen any longer and allowed me to realize sideEffects was missing from the webhook. So I went back to the code to see if the error creating the webhook was misinterpreted as the error writing the certificate, and I don't see it. I am not complaining.
> 
> I am using a different email so I guess it needs approval to run CI. cc @zshi-redhat @martinkennelly

I did notice that sideEffects didn't work when I removed the securityContext but didn't realise why.

**Comment by Eoghan1232** [comment]
> @Eoghan1232 @martinkennelly would you mind take another look?

NRI no longer crashes and successfully started up.
However, the E2E tests fails for me locally. Not sure if it's a problem on my side. all 4 tests fails. 
@zshi-redhat @jcaamano 
```
• Failure [40.037 seconds]
Network injection testing
/root/workarea/network-resources-injector/test/e2e/httpserver_test.go:10
  one network request
  /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:14
    should have one limit injected [It]
    /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:28

    Expected
        <bool>: false
    to be true

```

**Comment by jcaamano** [comment]
> > @Eoghan1232 @martinkennelly would you mind take another look?
> 
> NRI no longer crashes and successfully started up.
> However, the E2E tests fails for me locally. Not sure if it's a problem on my side. all 4 tests fails.
> @zshi-redhat @jcaamano
> 
> ```
> • Failure [40.037 seconds]
> Network injection testing
> /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:10
>   one network request
>   /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:14
>     should have one limit injected [It]
>     /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:28
> 
>     Expected
>         <bool>: false
>     to be true
> ```

Did work for me locally and it's working now in this CI, hmmm :/

**Comment by zshi-redhat** [comment]
> > @Eoghan1232 @martinkennelly would you mind take another look?
> 
> NRI no longer crashes and successfully started up.
> However, the E2E tests fails for me locally. Not sure if it's a problem on my side. all 4 tests fails.
> @zshi-redhat @jcaamano
> 
> ```
> • Failure [40.037 seconds]
> Network injection testing
> /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:10
>   one network request
>   /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:14
>     should have one limit injected [It]
>     /root/workarea/network-resources-injector/test/e2e/httpserver_test.go:28
> 
>     Expected
>         <bool>: false
>     to be true
> ```

I tried to verify it locally but have trouble pulling the kind image from docker, always hit rate limit. How did you workaround it?

**Comment by Eoghan1232** [comment]
Update: 
The e2e tests now pass locally for me.
My issue was that I did not have "intel_iommu=on" in my grub file. not sure why but alas, it passed for me locally now :) 
@jcaamano @zshi-redhat 

I also have not faced any issues pulling the kind image from docker, a workaround might be to pull the image beforehand and point the pod spec, etc at the image.
 
This seems good to go from me!

**Comment by zshi-redhat** [comment]
@Eoghan1232 can you approve (we need two approves)? then I will merge it.

**Comment by Eoghan1232** [comment]
> @Eoghan1232 can you approve (we need two approves)? then I will merge it.

Unfortunately, don't have the rights to approve. Will need to look into getting maintainer rights in the future.
@zshi-redhat

**Comment by zshi-redhat** [comment]
> > @Eoghan1232 can you approve (we need two approves)? then I will merge it.
> 
> Unfortunately, don't have the rights to approve. Should I look into getting maintainer rights for future?
> @zshi-redhat

Yes, please if you'd like to commit to this or other related projects.

**Comment by zshi-redhat** [comment]
I'm going to merge this PR since it has been verified by Eoghan1232 and jcaamano locally, and CI passes.

**Pull request merged**
